crates/corro-agent/src/agent/util.rs

```diff
@@ -894,23 +942,100 @@ pub async fn process_multiple_changes(
 
+        let mut last_cleared: Option<Timestamp> = None;
         for (actor_id, versions_set) in overwritten {
+            if actor_id != agent.actor_id() {
+                warn!("clearing and setting timestamp for empties from a different node");
+            }
             for versions in versions_set {
-                store_empty_changeset(&tx, actor_id, versions)?;
+                let ts = Timestamp::from(agent.clock().new_timestamp());
+                let inserted = store_empty_changeset(&tx, actor_id, versions, ts)?;
+                if inserted > 0 {
+                    last_cleared = Some(ts);
+                }
             }
         }
 
+        for (actor_id, ts_versions) in &sent_empties {
+            println!("handling empties in sent_empties");
+
+            let mut max_ts = Default::default();
+            for (versions, ts) in ts_versions.iter() {
+                for version in versions {
+                    store_empty_changeset(&tx, *actor_id, version.clone(), *ts)?;
+                    if *ts > max_ts {
+                        max_ts = *ts;
+                    }
+                }
+            }
+
+            let mut snap = match snapshots.remove(&actor_id) {
+                Some(snap) => snap,
+                None => {
+                    let booked = {
+                        bookie
+                            .blocking_write(format!(
+                                "process_multiple_changes(for_actor_blocking):{actor_id}",
+                            ))
+                            .ensure(*actor_id)
+                    };
+
+                    let booked_write = booked.blocking_write(format!(
+                        "process_multiple_changes(booked writer, handling requested empties):{actor_id}",
+                    ));
+                    booked_write.snapshot()
+                }
+            };
+
+            snap.update_cleared_ts(&tx, max_ts)
+                .map_err(|source| ChangeError::Rusqlite {
+                    source,
+                    actor_id: Some(*actor_id),
+                    version: None,
+                })?;
+
+            snap.insert_db(
+                &tx,
+                ts_versions.iter().flat_map(|s| s.0.iter()).fold(
+                    RangeInclusiveSet::new(),
+                    |mut acc, versions| {
+                        acc.insert(versions.clone());
+                        acc
+                    },
+                ),
+            )
+            .map_err(|source| ChangeError::Rusqlite {
+                source,
+                actor_id: Some(*actor_id),
+                version: None,
+            })?;
+            snapshots.insert(*actor_id, snap);
+        }
+
+        let mut booked_writer = agent
+            .booked()
+            .blocking_write("process_multiple_changes(update_cleared_ts)");
+        let mut snap = booked_writer.snapshot();
+        if let Some(ts) = last_cleared {
+            snap.update_cleared_ts(&tx, ts)
```

Comment:
🤔 isn't that redundant?

Reply:
the one above if for the received empties. this one is for the node's own empties (since we check for the node's overwrittten versions and we clear them in this function)